docs/container-affinity.md

```diff
@@ -0,0 +1,225 @@
+push or the pull is. This is useful in situations where the policy needs to make some
+compromises because an optimal placement is not possible. The weight then also acts as
+a way to specify preferences of priorities between the various compromises: the heavier
+the weight the stronger the pull or push and the larger the propbability that it will be
+honored, if this is possible at all.
+
+The scope can be omitted from an affinity in which case it implies *Pod scope*, IOW
+the scope of all containers that belong to the same Pod as the container for which
+which the affinity is defined.
+
+The weight can also be omitted in which case it defaults to -1 for anti-affinities
+and +1 for affinities.
+
+Both the affinity scope and the expression select containers, therefore they are identical
+in nature. In fact, the scope is just another expression. An expression consists of
+three parts:
+
+  - key: specifies what *metadata* to pick from a container for evaluation
+  - operation (op): specifies what *logical operation* the expression evaluates
+  - values: a set of *strings* to evaluate the the value of the key against
+
+Essentially an expression defines a logical operation of the form (key op values).
+Evaluating this logical expression will take the value of the key in  which
+either evaluates to true or false. 
+a boolean true/false result. Currently the following operations are supported:
+
+  - Equals: equality, true if the *value of key* equals the single item in *values*
+  - NotEqual: inequality, true if the *value of key* is not equal to the single item in *values*
+  - In: membership, true if *value of key* equals to any among *values*
+  - NotIn: negated membership, true if the *value of key* is not equal to any among *values*
+  - Exists: true if the given *key* exists with any value
+  - NotExists: true if the given *key* does not exist
+
+
+The effective affinity between containers C_1 and C_1, A(C_1, C_2) is the sum of the
+weights of all pairwise in-scope matching affinities W(C_1, C_2). IOW, evaluating an
+affinity for a container C_1 is done by first using the scope (expression) to determine
+which containers are in the scope of the affinity. Then, for each in-scope container C_2
+for which the match expression evaluates to true, taking the weight of the affinity and
+adding it to the effective affinity A(C_1, C_2).
+
+Note that currently (for the topology-aware policy) this evaluation is asymmetric:
+A(C_1, C_2) and A(C_2, C_1) can and will be different unless the affinity annotations
+crafted to prevent this (IOW to be symmetric). Moreover, A(C_1, C_2) is calculated
+and taken into consideration during resource allocation for C_1, while A(C_2, C_1)
+is calculated and taken into account during resource allocation for C_2. This might be
+changed in a future version.
+
+
+## Examples
+
+Put the container `peter` close to the container `sheeps` but far away from the
```

Comment:
humorous nitpick.... the plural of sheep... is sheep (not sheeps :-) )

Reply:
Ah, thanks. If only Shaun the Sheep had a proper spoken audio track... with so many hours under my belt I might have realized this then. ;-)